bitsandbytes/backends/base.py

```diff
@@ -0,0 +1,133 @@
+from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+
+import torch
+
+from bitsandbytes.utils import QuantState
+
+
+class Backend(ABC):
```

Comment:
We cannot use an abstract base class here. This makes the interface too big to implement. We want people to be able to contribute sub-interfaces, for example, only implement the 4-bit functionality but not the 8-bit and 8-bit optimizer functionality. The intent of such a design it better captured by a base class that implements these functions with an `NotImplementedError` exception which have to be overridden by the backend.

I think the intend would be even clearer by having 4 backends: 4-bit, 8-bit, 8-bit optimizers, block-wise quantization. However, this will also introduce more bloat in terms of boilerplate and more classes. Not sure how to handle this and feedback would be appreciated. I think it might be better to have a single class and just highlight both as comments and in the documentation that not all functions need to be overridden for a solid contribution.

Reply:
I see now that the methods already throw a NotImplementedError. I think this is good already. So just removing the ABC would make it possible to implement sub-interfaces. I think to make the sub-interfaces clearer it would be great to have a NotImplementedError that shows the set of functions that need to be implemented. For example,

```python
mm_dequant(...)
...
raise NotImplementedError("mm_dequant not implemented! \
This function is part of the 8-bit interface and it needs to be implemented along with: \
mm_dequant, igemmlt, extract_outliers, double_quant")
```

Reply:
If it's okay to partially implement a backend, then sure, we can make it a concrete base class with NotImplementeds thrown around.

Reply:
Shouldn't partial implementations fall back to CPU?